src/emulator/storage/metadata.ts

```diff
@@ -30,35 +39,51 @@ export class StoredFileMetadata {
   customMetadata: { [s: string]: string };
 
   constructor(
-    bucketId: string,
-    objectId: string,
     bytes: Buffer,
-    contentType: string,
-    contentEncoding: string | undefined,
-    incomingMetadata: IncomingMetadata,
+    opts: Partial<SerializedFileMetadata> & {
```

Comment:
I refactored this constructor to make it possible to rehydrate metadata from serialized JSON without side effects. I think it's actually a lot cleaner now.